Plug log agent kubernetes integration to autodiscovery

achntrl opened the pull request:
### What does this PR do?

- Don't send sources and services to channels if there are no consumers instead of always consuming.
- Cleaned launchers.
- Added unit-tests
- Refactor the logic of k8 integration in log agent to use autodiscovery

### Motivation

Today, the logs-scheduler can be blocking and freeze the autodiscovery if a new config is scheduled and no component consumes it.
Unify the logic in the log agent to only use AD and remove pod watchers

### Additional Notes

This PR is a rebase of #2326 on top of #2324

achntrl committed:
Plug k8 to the scheduler output

achntrl committed:
Remove pod watcher

achntrl committed:
Add release note

achntrl committed:
Unify constants

achntrl committed:
Inject the right sources at construction

In kubernetes integration

achntrl committed:
Fix copy-paste bug

achntrl committed:
Fix getting pods that were already deleted

achntrl committed:
Remove double logging

achntrl committed:
Rename k8 scanner to launcher

achntrl committed:
Rename variables

achntrl committed:
Enable containerd

ajacquemot commented on pkg/logs/config/integration_config.go:
```diff
@@ -8,14 +8,17 @@ package config
 	"fmt"
 	"regexp"
+
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
```
This is not needed as we control the configuration format here.

ajacquemot commented on pkg/logs/config/integration_config.go:
```diff
@@ -8,14 +8,17 @@ package config
 	FileType         = "file"
-	DockerType       = "docker"
+	DockerType       = containers.RuntimeNameDocker
+	ContainerdType   = containers.RuntimeNameContainerd
```
There is no config for `containerd` yet and I don't think we'll need one since we'd like to deprecate `docker`.

ajacquemot commented on pkg/logs/input/container/scanner.go:
```diff
@@ -33,7 +33,7 @@ func NewLauncher(sources *config.LogSources, services *service.Services, pipelin
 		}
 		// attempt to initialize a kubernetes scanner
 		log.Warnf("Could not setup the docker scanner, falling back to the kubernetes one: %v", err)
-		scanner, err := kubernetes.NewScanner(sources, services)
```
👍

ajacquemot commented on pkg/logs/service/providers.go:
```diff
@@ -5,7 +5,12 @@
 package service
 
+import (
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
```
👍

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,219 @@
+}
+
+// getAnnotation returns the logs-config annotation for container if present.
+func (l *Launcher) getAnnotation(pod *kubelet.Pod, container kubelet.ContainerStatus) string {
```
Okay for keeping the annotation here for now but I think we should add a FIXME and create a card to remove this logic as it's done in `autodiscovery`.

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,219 @@
+			l.addSources(service)
+		case service := <-l.dockerRemovedServices:
+			l.removeSources(service)
+		case <-l.dockerSources:
```
do we really want to keep the `dockerSources` case ?

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,219 @@
+			l.addSources(service)
+		case service := <-l.containerdRemovedServices:
+			l.removeSources(service)
+		case <-l.containerdSources:
```
dito: don't think we need to keep this case either

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,219 @@
+// addSourcesFromPod creates new log-sources for each container of the pod.
+// it checks if the sources already exist to avoid tailing twice the same
+// container when pods have multiple containers
+func (l *Launcher) addSourcesFromPod(pod *kubelet.Pod) {
```
nits `addSourcesForPod`

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,219 @@
+
+// addSources creates a new log-source from a service by resolving the
+// pod linked to the entityID of the service
+func (l *Launcher) addSources(service *service.Service) {
```
Should this method be renamed into `addSource` and just create a new file source for the `service.Identifier` ? This would enable us to get rid of the map but we'd have to navigate in the pod structure multiple times, dunno what's the best solution. WDYT ?

achntrl replied:
I think it makes no difference: 
Here, you get a container, you add a source for all container in its pod, then you receive the other containers of the same pod and you do nothing

Or you get a container, you have to find the matching one in the pod and add the source

I think the current logic is simpler since you don't have to handle edge cases

ajacquemot replied:
You're right the behaviour does not change but this method is meant to create a new source for a given service, here it's not what it does, instead it maps one service to a set of sources which is not semantically correct IMO. I'd rather make it clear mapping one service to one source so that we can get rid of the map.

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,219 @@
+}
+
+// removeSources removes a new log-source from a service
+func (l *Launcher) removeSources(service *service.Service) {
```
nits `removeSources` ->`removeSource`

ajacquemot reviewed (commented):
Happy to see that we removed some duplicated logic here, good job.
Left few comments.
Thanks

achntrl committed:
Remove sources from launcher

Because they are resolved from the services

achntrl committed:
nits

achntrl committed:
Log when k8 launcher exits

achntrl committed:
Modify add source logic

achntrl committed:
Fix a channel bug between docker and k8

When docker setup fail, a channel has been created for sources and
the scheduler tries to push things into sources, but k8 integration
won't read it so the scheduler is locked

achntrl committed:
Refactor addSource

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,220 @@
+	}
+	container, err := l.searchContainer(service, pod)
+	if err != nil {
+		log.Debug(err)
```
I think we should return here and log a warning

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -0,0 +1,220 @@
+		if service.GetEntityID() == containerID {
+			// If the container is already tailed, we don't do anything
+			// That shoudn't happen
+			if _, exists := l.sourcesByContainer[containerID]; exists {
```
I don't think it should be the responsibility of the this method to assert that the source has already been processed. I would move this upstream, i.e in `addSource`.

achntrl committed:
Move already processed source logic in addSource

ajacquemot reviewed (approved):
LGTM, thanks for tackling all the feedback, I'd like to someone to give another 👍 though before merging please
